src/amd/compiler/aco_lower_bool_phis.cpp

```diff
@@ -206,34 +207,27 @@ void lower_linear_bool_phi(Program *program, Block *block, aco_ptr<Instruction>&
 void lower_bool_phis(Program* program)
 {
    for (Block& block : program->blocks) {
-      std::vector<aco_ptr<Instruction>> instructions;
-      std::vector<aco_ptr<Instruction>> non_phi;
-      instructions.swap(block.instructions);
-      block.instructions.reserve(instructions.size());
-      unsigned i = 0;
-      for (; i < instructions.size(); i++)
-      {
-         aco_ptr<Instruction>& phi = instructions[i];
-         if (phi->opcode != aco_opcode::p_phi && phi->opcode != aco_opcode::p_linear_phi)
-            break;
-         if (phi->opcode == aco_opcode::p_phi && phi->definitions[0].regClass() == s2) {
-            non_phi.emplace_back(std::move(lower_divergent_bool_phi(program, &block, phi)));
-         } else if (phi->opcode == aco_opcode::p_linear_phi && phi->definitions[0].regClass() == s1) {
+      /* find the first non phi instruction */
+      auto IsPhi = [] (aco_ptr<Instruction>& phi) -> bool { return is_phi(phi); };
+      auto end = std::find_if_not(block.instructions.begin(), block.instructions.end(), IsPhi);
+
+      /* partition divergent boolean phi's to the back */
+      auto IsNotDivergentBooleanPhi = [] (aco_ptr<Instruction>& phi) -> bool {
+         return !(phi->opcode == aco_opcode::p_phi && phi->definitions[0].regClass() == s2);
+      };
+      std::stable_partition(block.instructions.begin(), end, IsNotDivergentBooleanPhi);
+
+      /* lower the boolean phi's */
+      const size_t endSize = std::distance(block.instructions.begin(), end);
+      for (size_t i = 0; i < endSize; ++i) {
+         aco_ptr<Instruction>& phi = block.instructions[i];
+         if (phi->opcode == aco_opcode::p_linear_phi && phi->definitions[0].regClass() == s1) {
             /* if it's a valid non-boolean phi, this should be a no-op */
             lower_linear_bool_phi(program, &block, phi);
-            block.instructions.emplace_back(std::move(phi));
-         } else {
-            block.instructions.emplace_back(std::move(phi));
          }
-      }
-      for (auto&& instr : non_phi) {
-         assert(instr->opcode != aco_opcode::p_phi && instr->opcode != aco_opcode::p_linear_phi);
-         block.instructions.emplace_back(std::move(instr));
-      }
-      for (; i < instructions.size(); i++) {
-         aco_ptr<Instruction> instr = std::move(instructions[i]);
-         assert(instr->opcode != aco_opcode::p_phi && instr->opcode != aco_opcode::p_linear_phi);
-         block.instructions.emplace_back(std::move(instr));
+         else if (!IsNotDivergentBooleanPhi(phi)) {
+            phi = lower_divergent_bool_phi(program, &block, phi);
```

Comment:
Since this often will (but not always) add an instruction in the beginning of `block.instructions`, `i` and `endSize` will need to be adjusted. I think just comparing `block.instructions.size()` before and after `lower_divergent_bool_phi()` will be fine IMO.

This also has an issue where `phi` points to freed memory because the reference is not longer valid, so `phi` will have to be passed as an `Instruction *` (`std::move`'ing it might cause issues with `update_phi()`) and it will have to be updated by indexing `block.instructions` again.

Reply:
Hm I have to think about the later part a bit the original code is:
```cpp
    non_phi.emplace_back(std::move(lower_divergent_bool_phi(program, &block, phi)))
```
Now we directly overwrite the old instruction with the result from `lower_divergent_bool_phi`.
```cpp
    phi = lower_divergent_bool_phi(program, &block, phi);
```

Do you say that we should move it?
```cpp
    phi = std::move(lower_divergent_bool_phi(program, &block, phi));
```

I believe it would be better if we would use `reset`
```cpp
    phi.reset(lower_divergent_bool_phi(program, &block, phi));
```

Reply:
I agree on reset although it should be essentially the same.
@pendingchaos sure you aren't confusing things? An Instruction* gets invalidated while aco_ptr<Instruction>& should stay valid if you just swap out the content.

Reply:
@daniel-schuermann : The `aco_ptr&` is a reference to a vector element, so the reference is invalidated once a new element is inserted.

@miscco : how could reset() be used? lower_divergent_bool_phi() returns a unique_ptr. I think it should be directly overwritten, otherwise std::move() used if directly overwriting doesn't compile.

So something like:
```c++
      block.instructions[i] = lower_divergent_bool_phi(program, &block, phi.get());
```
using the adjusted `i` and `lower_divergent_bool_phi` modified to take an `Instruction *`

Reply:
> @daniel-schuermann : The `aco_ptr&` is a reference to a vector element, so the reference is invalidated once a new element is inserted.

Woops.
> @miscco : how could reset() be used? lower_divergent_bool_phi() returns a unique_ptr. I think it should be directly overwritten, otherwise std::move() used if directly overwriting doesn't compile.
> 
> So something like:
> 
> ```c++
>       block.instructions[i] = lower_divergent_bool_phi(program, &block, phi.get());
> ```
> 
> using the adjusted `i` and `lower_divergent_bool_phi` modified to take an `Instruction *`
Ahh, I thought you want to use the pointer in the other direction.
You can do block.instructions[i].reset(lower_..().release()), but that is essentially the same.

Reply:
Ah yeah, unique_ptr::release() is a thing. With that, the function can return an Instruction *. Either way is fine, I think

Reply:
I had a closer look at `lower_divergent_bool_phi` and I believe the best way forward is to simply create a temporary buffer. 

Note that this would most likely be more efficient as currently we search (possibly multiple times) for the same insertion point over and over in `insert_before_logical_end`.

So my suggestion would be to change the code so that `lower_divergent_bool_phi` takes a reference to the divergent boolean phi instruction and rewrites it. All possible additional instruction are collected in an temporary buffer and then inserted once.

Reply:
I just realized that we actually only ever insert instructions into other blocks, namely the logical predecessors of the operand (`logical_preds`). So the actual instruction vector of the given block we are currently handling remains unchanged and we should not need to worry about iterator invalidation. 

Note that this would otherwise also affect `lower_linear_bool_phi`

Reply:
I re-checked it and unfortunately it does not only insert into predecessors. The returned s_mov_b64 gets it's Operand from get_ssa() which creates linear phis for diverged names (I think that is always the case on the current block).
Not really sure how to work around that. In theory, if we insert a linear_phi, we don't really need the s_mov. So, we could return the linear_phi instead of the s_mov or even change it in-place (which should be most efficient). To do that I'd suppose copying the else-part from get_ssa into lower_divergent_bool_ph() and not returning a new instruction but changing the existing phi.